jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -286,32 +273,56 @@ public int read(byte[] b, int off, int len) throws IOException
                 }
             }
 
+            long timeout = isAsync()?-1:getHttpChannelState().getHttpChannel().getBlockingTimeout();
+
             // Consume content looking for bytes to read
             while (true)
             {
-                Content item = nextContent();
-                if (item != null)
+                long start = timeout>0?System.nanoTime():0;
+                
+                try
                 {
-                    l = get(item,b,off,len);
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} read {} from {}",this,l,item);
-
-                    // Consume any following poison pills
-                    if (item.isEmpty())
-                        nextInterceptedContent();
-                    break;
+                    Content item = nextContent();
+                    if (item != null)
+                    {
+                        l = get(item,b,off,len);
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("{} read {} from {}",this,l,item);
+
+                        // Consume any following poison pills
+                        if (item.isEmpty())
+                            nextInterceptedContent();
+                        break;
+                    }
+                    
+                    // Update and check the budget before blocking
+                    if (timeout>0)
```

Comment:
@sbordet can you give this bit a quick review?   I now update and check the budget after doing the reading and before blocking.